core/src/main/java/org/ldaptive/url/Url.java

```diff
@@ -0,0 +1,443 @@
+      }
+      return defaultPort;
+    }
+    return port;
+  }
+
+
+  /**
+   * Returns false if a port was supplied in this url.
+   *
+   * @return  false if a port was supplied in this url
+   */
+  public boolean isDefaultPort()
+  {
+    return port == -1;
+  }
+
+
+  /**
+   * Returns the base DN.
+   *
+   * @return  baseDn
+   */
+  public String getBaseDn()
+  {
+    return baseDn == null ? DEFAULT_BASE_DN : baseDn;
+  }
+
+
+  /**
+   * Returns the parsed base DN.
+   *
+   * @return  parsed baseDn
+   */
+  public Dn getParsedBaseDn()
+  {
+    return parsedBaseDn == null ? DEFAULT_PARSED_BASE_DN : parsedBaseDn;
+  }
+
+
+  /**
+   * Returns whether a base DN was supplied in this url.
+   *
+   * @return  whether a base DN was supplied in this url
+   */
+  public boolean isDefaultBaseDn()
+  {
+    return baseDn == null;
+  }
+
+
+  /**
+   * Returns the attributes.
+   *
+   * @return  attributes
+   */
+  public String[] getAttributes()
+  {
+    return attributes == null ? DEFAULT_ATTRIBUTES : attributes;
+  }
+
+
+  /**
+   * Returns whether attributes were supplied in this url.
+   *
+   * @return  whether an attributes were supplied in this url
+   */
+  public boolean isDefaultAttributes()
+  {
+    return attributes == null;
+  }
+
+
+  /**
+   * Returns the scope.
+   *
+   * @return  scope
+   */
+  public SearchScope getScope()
+  {
+    return scope == null ? DEFAULT_SCOPE : scope;
+  }
+
+
+  /**
+   * Returns whether a scope was supplied in this url.
+   *
+   * @return  whether a scope was supplied in this url
+   */
+  public boolean isDefaultScope()
+  {
+    return scope == null;
+  }
+
+
+  /**
+   * Returns the filter.
+   *
+   * @return  filter
+   */
+  public String getFilter()
+  {
+    return filter == null ? DEFAULT_FILTER : filter;
+  }
+
+
+  /**
+   * Returns the parsed filter.
+   *
+   * @return  parsed filter
+   */
+  public Filter getParsedFilter()
+  {
+    return parsedFilter == null ? DEFAULT_PARSED_FILTER : parsedFilter;
+  }
+
+
+  /**
+   * Returns whether a filter was supplied in this url.
+   *
+   * @return  whether a filter was supplied in this url
+   */
+  public boolean isDefaultFilter()
+  {
+    return filter == null;
+  }
+
+
+  /**
+   * Returns a string representation of this LDAP URL. Uses a {@link MinimalUrlFormatter} by default.
+   *
+   * @return  string form of the LDAP URL
+   */
+  public String format()
+  {
+    return format(new MinimalUrlFormatter());
+  }
+
+
+  /**
+   * Returns a string representation of this LDAP URL.
+   *
+   * @param  formatter  to produce the string
+   *
+   * @return  string form of the LDAP URL
+   */
+  public String format(final UrlFormatter formatter)
+  {
+    return formatter.format(this);
+  }
+
+
+  /**
+   * Determines whether the supplied scheme is valid. Must be one of 'ldap', 'ldaps' or 'ldapi'.
+   *
+   * @param  scheme  to validate
+   */
+  static void validateScheme(final String scheme)
```

Comment:
I have a slight preference that these validations live in a utility class. I believe `LdapUtils` is a suitable existing location.

Reply:
These are very specific to the `Url` implementations, I'd prefer not to expose them generally until a general use case arises.